### ISSUE TITLE
_PythonScriptingInstanceInput Missing Methods

### ISSUE THREAD

**Issue by ehennenfent** [Vector35/binaryninja-api#769]
Consider this a formal submission of [@thebarbershopper's request on the slack](https://binaryninja.slackarchive.io/api-help/page-52/ts-1501976958539058)

While the TTY issues with iPython mentioned in #392 have been fixed, other programs still break as a result of Binary Ninja's stdin/stdout shenanigans. For example, consider the following behavior, observed with version 1.0.862-dev

```
>>> import binaryninja 
>>> from prompt_toolkit import prompt
>>> prompt(u'test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/prompt_toolkit/shortcuts.py", line 548, in prompt
    eventloop=eventloop)
  File "/usr/local/lib/python2.7/site-packages/prompt_toolkit/shortcuts.py", line 576, in run_application
    output=create_output(true_color=true_color))
  File "/usr/local/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 81, in __init__
    self.input = input or StdinInput(sys.stdin)
  File "/usr/local/lib/python2.7/site-packages/prompt_toolkit/input.py", line 67, in __init__
    assert self.stdin.isatty()
AttributeError: '_PythonScriptingInstanceInput' object has no attribute 'isatty'
```

Upon applying the fix suggested via [this comment on #392](https://github.com/Vector35/binaryninja-api/issues/392#issuecomment-263593219), the traceback above is resolved.

**Comment by psifertex** [comment]
Unfortunately that patch broke other things in testing though I don't have it documented right now which ones it was. I'll leave this open for now until I can document it.

**Comment by psifertex** [comment]
Did some more testing -- logging is broken in all internal functions (or any external ones too using the log module) using the comment from the other issue. You're welcome to choose which broken behavior is more acceptable to you (built in logging failing or fake terminal causing trouble with other code that relies on a tty), but until we can work out a fix for both we're probably going to leave it the way it is for now.

**Comment by ehennenfent** [comment]
In this case, it's enough to mess with the constructor for `_PythonScriptingInstanceInput`
```
class _PythonScriptingInstanceInput(object):
	def __init__(self, orig):
		self.orig = orig
		self.isatty = lambda: True
		self.fileno = lambda: 0
```
While this fixes `prompt_toolkit`, in the long run, it will probably be necessary to implement method stubs for all the methods that `_PythonScriptingInstanceInput` is missing.

**Comment by ehennenfent** [comment]
Seeing as both `_PythonScriptingInstanceInput` and `_PythonScriptingInstanceOutput` have a member called `orig` that points to the original stdin/stdout/stderr, one can add a custom `__getattr__` to each class that passes any missing methods down to the original object.
```
def __getattr__(self, name):
    return getattr(self.orig, name)
```

Note that unlike `__getattribute__`, `__getattr__` is only run if the object does not have the attribute already defined, so this will not overwrite the existing methods on the classes.

**Comment by psifertex** [comment]
fwiw -- my response still stands, just double checked, and implementing those changes cause all logging functions to be broken in headless scripts. I'm sure there's a way to fix it, but don't have the time to investigate now.

**Comment by cetfor** [comment]
I also ran into this issue with Binary Ninja's sys.std* shenanigans. It also seems like other people are running into [problems with Manticore and Binary Ninja](https://github.com/trailofbits/manticore/issues/199) too.

I'm using Python's multiprocessing library to handle headless processing of BNIL from very large binaries by splitting analysis of each function into multiple processes.  Without [this suggested fix](https://github.com/Vector35/binaryninja-api/issues/392#issuecomment-263593219) applied to `scriptingprovider.py` I getting similar issues everyone else is having.

```Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 249, in _bootstrap
    sys.stdin.close()
AttributeError: '_PythonScriptingInstanceInput' object has no attribute 'close'
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 249, in _bootstrap
    sys.stdin.close()
AttributeError: '_PythonScriptingInstanceInput' object has no attribute 'close'
```

The fix linked above works for us internally. It's a shame it breaks other functionality since it's impractical for us to push this code for others to use and require them to modify their python bindings.

**Comment by psifertex** [comment]
We've implemented a change so that headless plugins will work as of versions >1200. 

Unfortunately, our python console is NOT a full terminal. A full fix will require #30 to be resolved inside the UI itself, but this issue should be resolved for headless terminals/scripts at least.
